packages/ts/react-signals/src/FullStackSignal.ts

```diff
@@ -0,0 +1,167 @@
+import { signal } from '@preact/signals-react';
+import type { ConnectClient, Subscription } from '@vaadin/hilla-frontend';
+import { nanoid } from 'nanoid';
+import { Signal } from './core.js';
+
+const ENDPOINT = 'SignalsHandler';
+
+type SubscriptionEvent<T> = Readonly<{
+  id: string;
+  type: StateEventType;
+  value: T;
+}>;
+
+/**
+ * Types of changes that can be produced or processed by a signal.
+ */
+export enum StateEventType {
+  SET = 'set',
+  SNAPSHOT = 'snapshot',
+}
+
+/**
+ * An object that describes the change of the signal state.
+ */
+export type StateEvent<T> = Readonly<{
+  id: string;
+  type: StateEventType;
+  value: T;
+}>;
+/**
+ * An object that describes a data object to connect to the signal provider
+ * service.
+ */
+export type ConnectionData = Readonly<{
+  /**
+   * The client instance to be used for communication.
+   */
+  client: ConnectClient;
+
+  /**
+   * The name of the signal provider service endpoint.
+   */
+  endpoint: string;
+
+  /**
+   * The name of the signal provider service method.
+   */
+  method: string;
+}>;
+
+class SubscriptionManager<T> {
+  readonly #id: string;
+  readonly #data: ConnectionData;
+  #subscription?: Subscription<StateEvent<T>>;
+
+  constructor(id: string, data: ConnectionData) {
+    this.#data = data;
+    this.#id = id;
+  }
+
+  get subscription() {
+    return this.#subscription;
+  }
+
+  subscribe() {
+    const { client, endpoint, method } = this.#data;
+
+    this.#subscription ??= client.subscribe(ENDPOINT, 'subscribe', {
+      providerEndpoint: endpoint,
+      providerMethod: method,
+      clientSignalId: this.#id,
+    });
+
+    return this.#subscription;
+  }
+
+  async call(event: SubscriptionEvent<T>): Promise<void> {
+    await this.#data.client.call(ENDPOINT, 'update', {
+      clientSignalId: this.#id,
+      event,
+    });
+  }
+
+  cancel() {
+    this.#subscription?.cancel();
+    this.#subscription = undefined;
+  }
+}
+
+/**
+ * A signal that holds a shared value. Each change to the value is propagated to
+ * the server-side signal provider. At the same time, each change received from
+ * the server-side signal provider is propagated to the local signal and it's
+ * subscribers.
+ *
+ * @internal
+ */
+export abstract class FullStackSignal<T> extends Signal<T> {
+  readonly id = nanoid();
+  readonly #pending = signal(false);
+  readonly #error = signal<Error | undefined>(undefined);
+  readonly #manager: SubscriptionManager<T>;
+
+  constructor(value: T | undefined, data: ConnectionData) {
```

Comment:
Naming of the second parameter variable and type (`data: ConnectionData`) seems to be a bit unfortunate, specially when it is used alongside with `value` (first param). I suggest changing it to something that announces its purpose, such as `connectionConfig` or something else.